### PR TITLE
[SYCL][FPGA] Support Intel FPGA simulator device selector

### PR DESCRIPTION
**Pull request by mendell27**
Create **fpga_simulator_selector** to be used to select the FPGA simulator device. This assumes that the user has compiled their program with -Xssimulation (as well as -fintelfpga) to prepare for the simulator.

An object of class fpga_simulator_selector must be constructed early in the execution of the host program in order to make the simulator available at runtime.

The current implementation has the effect that if an object of class sycl::ext::intel::fpga_simulator_selector is defined, FPGA hardware devices selected using sycl::ext::intel::fpga_selector will select simulator devices. This will be documented in the release notes and the specification.  We expect this behaviour to be eliminated in the future.

### CODE REVIEW

**Comment by mendell27** [comment]
The failing tests have nothing to do with my change.  What can/should I do?

**Comment by mendell27** [comment]
don't approve yet.  still working on Windows support

**Comment by mendell27** [comment]
I changed my mind.  This is now okay to review.  We will live with the hardware/simulator restriction as if is unlikely for a user to want both simultaneously.

I have updated to re-use existing code

**Comment by mendell27** [comment]
How do I fix a failure in a clang file that I didn't change? clang/lib/Sema/SemaChecking.cpp

File: sycl/include/sycl/ext/intel/fpga_device_selector.hpp [hunk 1]
```diff
@@ -55,6 +59,22 @@ class fpga_emulator_selector : public platform_selector {
...
+  fpga_simulator_selector(unsigned int numDevices = 1) {
+    // Tell the runtime to start a simulator device
+    // Notes:
+    //   This replaces any hardware devices present
+    //   Currently only 1 simulator device is supported
+    auto devices = std::to_string(numDevices);
+#ifdef _WIN32
+    _putenv_s("CL_CONTEXT_MPSIM_DEVICE_INTELFPGA", devices.c_str());
+#else
+    setenv("CL_CONTEXT_MPSIM_DEVICE_INTELFPGA", devices.c_str(), 0);
```
Comment (romanovvlad) [inline sycl/include/sycl/ext/intel/fpga_device_selector.hpp:73]: Can we go further and just recommend users to set `CL_CONTEXT_MPSIM_DEVICE_INTELFPGA` env var before running their app to target simulator? In this case they would not need to rewrite and recompile the code to replace `fpga_selector` with `fpga_simulator_selector` ?

**Review by romanovvlad — CHANGES_REQUESTED** [review]

File: sycl/include/sycl/ext/intel/fpga_device_selector.hpp [hunk 2]
```diff
@@ -55,6 +59,22 @@ class fpga_emulator_selector : public platform_selector {
...
+class fpga_simulator_selector : public fpga_selector {
+public:
+  fpga_simulator_selector(unsigned int numDevices = 1) {
+    // Tell the runtime to start a simulator device
+    // Notes:
+    //   This replaces any hardware devices present
+    //   Currently only 1 simulator device is supported
+    auto devices = std::to_string(numDevices);
+#ifdef _WIN32
+    _putenv_s("CL_CONTEXT_MPSIM_DEVICE_INTELFPGA", devices.c_str());
```
[the remark this answers is not in the document]
Comment (mendell27) [reply sycl/include/sycl/ext/intel/fpga_device_selector.hpp:71]: It is documented in [oneAPI](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/programming-interface/fpga-flow/evaluate-kernel-through-simulation.html) and for OpenCL in 11.1.7.2. Simulating Your OpenCL Library in [OpenCL](https://manuals.plus/m/a9b58a94add52160154a9e2821a6937c0c6d1336bfd8aa6c22da25721d04b514#iframe)

For oneAPI, the intent is to hide this environment variable from the user, as it is very intrusive. OpenCL has a multi-step process for running programs, while oneAPI hides all that.  We would prefer it not to be openly documented.

When compiling for OpenCL for simulation, the **aoc** command explicitly tells the user to set the variable.  For oneAPI, **aoc** does not print this message

File: sycl/include/sycl/ext/intel/fpga_device_selector.hpp [hunk 3]
```diff
@@ -55,6 +59,22 @@ class fpga_emulator_selector : public platform_selector {
   fpga_emulator_selector() : platform_selector(EMULATION_PLATFORM_NAME) {}
 };
 
+class fpga_simulator_selector : public fpga_selector {
+public:
+  fpga_simulator_selector(unsigned int numDevices = 1) {
+    // Tell the runtime to start a simulator device
+    // Notes:
+    //   This replaces any hardware devices present
+    //   Currently only 1 simulator device is supported
```
[the remark this answers is not in the document]
Comment (mendell27) [reply sycl/include/sycl/ext/intel/fpga_device_selector.hpp:68]: The restriction of one device is an ACL runtime issue, that might be changed in the future.  At that time, we can remove this comment.  The current documentation states that CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 is the way to simulate it.

**Comment by mendell27** [comment]
Is it possible to resolve this today?  We want to get the code sample changes in by CC (Thursday?), so there is a bit of a rush

File: sycl/include/sycl/feature_test.hpp.in [hunk 4]
```diff
@@ -62,7 +62,9 @@ __SYCL_INLINE_VER_NAMESPACE(_V1) {
 #define SYCL_EXT_INTEL_EXTENDED_ATOMICS 1
 #endif
 #endif
-#define SYCL_EXT_INTEL_FPGA_DEVICE_SELECTOR 1
+#ifndef SYCL_EXT_INTEL_FPGA_DEVICE_SELECTOR
```
Comment (mendell27) [inline sycl/include/sycl/feature_test.hpp.in:65]: Is the #if needed?

**Comment by gmlueck** [comment]
Do you mind using this PR to update the extension specification to use the official [template](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc)?  I know the existing spec does not follow that template, but this seems like a good opportunity to remove some technical debt.

The specification needs to be updated anyway to make it clear that the new `fpga_simulator_selector` is added in revision 2 of the spec.  This will be easier to do if it follows the template.

> Create fpga_simulator_selector to be used to select the FPGA simulator device. This assumes that the user has compiled their program with -Xssimulation (as well as -fintelfpga) to prepare for the simulator.
>
> An object of class fpga_simulator_selector must be constructed early in the execution of the host program in order to make the simulator available at runtime.
>
> The emulator and hardware FPGA device selectors will prefer the emulator/hardware over a simulation device if both have the same device name.
>
> The current implementation has the effect that if an object of class sycl::ext::intel::fpga_simulator_selector is defined, FPGA hardware devices selected using sycl::ext::intel::fpga_selector will select simulator devices. This will be documented in the release notes. We expect this behaviour to be eliminated in the future.

Some of this information seems like important stuff that the user needs to know.  If so, it should be added to the extension specification, perhaps as a NOTE.  For example:

```
[NOTE]
====
In the DPC++ implementation, the application must be built with "-Xssimulation"
in order to use `sycl::ext::intel::fpga_simulator_selector`.
====
```

Note that the SYCL 2020 style for selectors is to define an *object* that is the selector, not to define a type that the application has to construct.  Do you plan to update this extension to follow that style?  Mentioning this now in case you want to migrate to the new style now and then add this new selector only with the new style.

File: sycl/include/sycl/feature_test.hpp.in (hunk 4 above)
Comment (gmlueck) [reply sycl/include/sycl/feature_test.hpp.in:65]: No, this is not right.  You should just have `#define SYCL_EXT_INTEL_FPGA_DEVICE_SELECTOR 2`

Comment (mendell27) [reply sycl/include/sycl/feature_test.hpp.in]: removed

File: sycl/include/sycl/ext/intel/fpga_device_selector.hpp [hunk 5]
```diff
@@ -34,6 +34,10 @@ class platform_selector : public device_selector {
     const platform &pf = device.get_platform();
     const std::string &platform_name = pf.get_info<info::platform::name>();
     if (platform_name == device_platform_name) {
+      const std::string name = device.get_info<sycl::info::device::name>();
+      // Prefer hardware devices over simulator ones.
+      if (name.find("SimulatorDevice") != std::string::npos)
```
[the remark this answers is not in the document]
Comment (mendell27) [reply sycl/include/sycl/ext/intel/fpga_device_selector.hpp]: removed

**Comment by mendell27** [comment]
Updated to the current template.  I am not going to change to the object format at this time, as the current implementation of the FPGA runtime needs to be enhanced to support having the simulator active at all times first

**Review by romanovvlad — DISMISSED** [review]

File: sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+sycl::queue deviceQueue{sycl::ext::intel::fpga_simulator_selector{}};
+....
+
+=== Select FPGA emulator device
+....
+// force FPGA emulator device
+sycl::queue deviceQueue{sycl::ext::intel::fpga_emulator_selector{}};
+....
+
+=== Feature test macro
```
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc:91]: Do you mind moving this sub-section to the beginning of "Specification"?  I know that the position of this sub-section is somewhat inconsistent across the specifications, but we are trying to standardize on making it the first sub-section under "Specification".

File: sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc [hunk 7]
```diff
@@ -0,0 +1,120 @@
+= sycl_ext_fpga_selector
```
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc:1]:
```suggestion
= sycl_ext_intel_fpga_device_selector
```

This should match the name of the extension.

File: sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+#include <sycl/ext/intel/fpga_device_selector.hpp>
+....
+to your source code.
+
+
+== Specification
+
+=== Select FPGA hardware device
+....
+// force FPGA hardware device
```
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc:75]:
```suggestion
// Select FPGA hardware device
```

Isn't it more appropriate to say "select" here (and below)?

File: sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+sycl::queue deviceQueue{sycl::ext::intel::fpga_selector{}};
+....
+
+=== Select FPGA simulator device
+....
+// force FPGA simulator device
+sycl::queue deviceQueue{sycl::ext::intel::fpga_simulator_selector{}};
+....
+
```
Comment (gmlueck) [inline sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc:84]: I'd add a sentence here like:

> Added in version 2 of this extension.

I know the feature-test table says this also, but readers are much more likely to notice the sentence here.

File: sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc (hunk 8 above)
Comment (mendell27) [reply sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc:75]: done

File: sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc (hunk 7 above)
Comment (mendell27) [reply sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc:1]: done

File: sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc (hunk 6 above)
Comment (mendell27) [reply sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc:91]: done

File: sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc (hunk 9 above)
Comment (mendell27) [reply sycl/doc/extensions/supported/sycl_ext_intel_fpga_device_selector.asciidoc:84]: done

**Review by gmlueck — APPROVED** [review]

File: sycl/include/sycl/ext/intel/fpga_device_selector.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (mendell27) [reply sycl/include/sycl/ext/intel/fpga_device_selector.hpp]: Do you have a problem with the code as it is now?

File: sycl/include/sycl/ext/intel/fpga_device_selector.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mendell27) [reply sycl/include/sycl/ext/intel/fpga_device_selector.hpp]: These comments have been removed

File: sycl/include/sycl/ext/intel/fpga_device_selector.hpp (hunk 1 above)
Comment (mendell27) [reply sycl/include/sycl/ext/intel/fpga_device_selector.hpp]: There are no plans for this current release to change this

**Review by romanovvlad — APPROVED** [review]

**Comment by mendell27** [comment]
Build check failures are not related to my changes

**Pull request merged**